File: 04_stack/02_min_stack/solution.py
"""
Stack -> Min Stack
Neetcode 22 / 150
Leetcode 155

Python

Comments:
Challenge is to track min even when it's popped. Take advantage of stack to have
a "history" by tracking the min at each push.

Time: O(1) duh that's the whole problem
Space: O(n) stack
"""


class MinStack:
    def __init__(self):
        # store val in first index, minimum at time in second index
        self.stack = []

    def push(self, val: int) -> None:
        past_min = self.getMin()
        if past_min is None or val < past_min:
            curr_min = val
        else:
            curr_min = past_min
        self.stack.append([val, curr_min])

    def pop(self) -> None:
        self.stack.pop()

    def top(self) -> int:
        if self.stack:
            return self.stack[-1][0]
        return None

    def getMin(self) -> int:
        if self.stack:
            return self.stack[-1][1]
        return None

File: 04_stack/02_min_stack/solution.js
/*
Stack -> Min Stack
Neetcode 22 / 150
Leetcode 155

Javascript

Comments:
Challenge is to track min even when it's popped. Take advantage of stack to have
a "history" by tracking the min at each push.

Time: O(1) duh that's the whole problem
Space: O(n) stack
*/

class MinStack {
	constructor() {
		this.stack = [];
	}

	/**
	 * @param {number} val
	 * @return {void}
	 */
	push(val) {
		let minVal = this.getMin();
		if (minVal === null || val < minVal) {
			minVal = val;
		}
		this.stack.push([val, minVal]);
	}

	/**
	 * @return {void}
	 */
	pop() {
		this.stack.pop();
	}

	/**
	 * @return {number}
	 */
	top() {
		if (this.stack.length) {
			return this.stack[this.stack.length - 1][0];
		}
		return null;
	}

	/**
	 * @return {number}
	 */
	getMin() {
		if (this.stack.length) {
			return this.stack[this.stack.length - 1][1];
		}
		return null;
	}
}
